File: Portfolio/Models/GameNight/GameNight.cs
﻿using Portfolio.Models.Auth;
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Linq;
using System.Threading.Tasks;

namespace Portfolio.Models.GameNight
{
    public class GameNight
    {
        [Key]
        public int Id { get; set; }

        [Required]
        public DateTime Date { get; set; }

        [ForeignKey("GameNightMeal")]
        public int? GameNightMealId { get; set; }

        public virtual GameNightMeal Meal { get; set; }

        [ForeignKey("User")]
        public string? UserId { get; set; }

        public virtual ApplicationUser User { get; set; }

        public List<GameNightGame> Games { get; set; }

        public bool? IsCancelled { get; set; }

        public List<GameNightUserStatus> UserStatuses { get; set; }

        public GameNight GetCancelledCopy()
        {
            return new GameNight { Date = Date, IsCancelled = true };
        }
    }
}

File: Portfolio/ClientApp/src/app/game-night/models/game-night.ts
import { User } from "src/app/auth/user";

export class GameNight {
    id: number;
    date: Date;
    userId?: string;
    gameNightMealId?: number;
    user?: User;
    meal?: GameNightMeal;
    games: GameNightGame[];
    isCancelled?: boolean;
    userStatuses: GameNightUserStatus[];
}

export class GameNightMeal {
    id: number;
    name: string;
    dateAdded: Date;
}

export class GameNightGame {
    id: number;
    name: string;
    image: string;
    minPlayers: number;
    maxPlayers: number;
}

export class GameNightUserStatus {
    id: number;
    gameNightId: number;
    userId: string;
    user: User;
    status: UserStatus;
}

export enum UserStatus {
    Available,
    Partial,
    NotAvailable,
    Unknown
}